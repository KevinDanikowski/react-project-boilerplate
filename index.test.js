// End-to-end Integration Test
const fs = require('fs')
const { execSync } = require('child_process')
const rimraf = require('rimraf')
const { error } = require('./helpers/logger')

const tmpDir = `tmp-${Math.floor(Math.random() * 1000000)}`

// TODO: Jest can't process coverage of spaned processes
// May need to wrap NYC to get the coverage of all the
// code executed here
// https://github.com/amclin/react-project-boilerplate/issues/28

describe('Integration Test', () => {
  describe('Generated App', () => {
    beforeAll(async () => {
      // Run the generator expecting successful STDOUT
      try {
        await execSync(
          `node ./index.js --use-npm --no-git --with-ssr -- ${tmpDir}`,
          { stdio: 'inherit' }
        )
      } catch (e) {
        error('Failed to complete generation process.', e)
        expect(true).toEqual(false) // Force test failure
      }
    })
    afterAll(() => {
      // Cleanup the temp
      rimraf.sync(`./${tmpDir}`, {}, () => {
        error(`No tmp directory (${tmpDir}) to remove.`)
        expect(true).toEqual(false) // Force test failure
      })
    })

    it('can build', async () => {
      // Generated app should be buildable
      try {
        await execSync(`(cd ${tmpDir} ; npm run build)`, { stdio: 'inherit' })
      } catch (e) {
        error('Generated app failed to build with `npm run build`', e)
        expect(true).toEqual(false) // Force test failure
      }
    })

    it('can export static html', async () => {
      // Generated app should be exportable
      try {
        await execSync(`(cd ${tmpDir} ; npm run export)`, { stdio: 'inherit' })
      } catch (e) {
        error('Generated app failed to export with `npm run export`', e)
        expect(true).toEqual(false) // Force test failure
      }
    })

    describe('New component wizard(plop)', () => {
      const componentTypes = {
        atom: {
          files: ['index.js', 'MockComponent.jsx', 'MockComponent.test.jsx']
        },
        molecule: {
          files: ['index.js', 'MockComponent.jsx', 'MockComponent.test.jsx']
        },
        organism: {
          files: ['index.js', 'MockComponent.jsx', 'MockComponent.test.jsx']
        }
      }

      // Loop through each type of component and test the files are created
      Object.entries(componentTypes).forEach(([componentType, data]) => {
        it(`can generate new ${componentType}s through the wizard`, async () => {
          const path = `${tmpDir}/src/components/${componentType}s/MockComponent`
          const cmd = `(cd ${tmpDir}; npm run generate -- component ${componentType} "mock component" "short description")`

          try {
            await execSync(cmd)
          } catch (e) {
            error(`Failed to run plop for a ${componentType}`, e)
            expect(true).toEqual(false) // Force test failure
          }

          // Check that all expected files are generated
          data.files.forEach(file => {
            expect(
              fs.readFileSync(`${path}/${file}`, 'utf-8')
            ).toMatchSnapshot()
          })
        })
      })

      it.skip('can generate new pages through the wizard', async () => {})
      it.skip('can generate new apis through the wizard', async () => {})
    })
  })
})
