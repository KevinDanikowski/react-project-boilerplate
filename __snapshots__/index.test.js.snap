// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Test Generated App New component wizard(plop) can generate new atoms through the wizard 1`] = `
"export { default } from './MockComponent'
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new atoms through the wizard 2`] = `
"import React from 'react'

const MockComponent = () => {
  return(
    <>
      <p>Boilerplate mock component component</p>
    </>
  )
}

export default MockComponent
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new atoms through the wizard 3`] = `
"import React from 'react'
import { render } from '@testing-library/react'
import MockComponent from '.'

describe('MockComponent', () => {
  it('creates short description', () => {
    const { asFragment } = render(<MockComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
})
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new molecules through the wizard 1`] = `
"export { default } from './MockComponent'
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new molecules through the wizard 2`] = `
"import React from 'react'

const MockComponent = () => {
  return(
    <>
      <p>Boilerplate mock component component</p>
    </>
  )
}

export default MockComponent
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new molecules through the wizard 3`] = `
"import React from 'react'
import { render } from '@testing-library/react'
import MockComponent from '.'

describe('MockComponent', () => {
  it('creates short description', () => {
    const { asFragment } = render(<MockComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
})
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new organisms through the wizard 1`] = `
"export { default } from './MockComponent'
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new organisms through the wizard 2`] = `
"import React from 'react'

const MockComponent = () => {
  return(
    <>
      <p>Boilerplate mock component component</p>
    </>
  )
}

export default MockComponent
"
`;

exports[`Integration Test Generated App New component wizard(plop) can generate new organisms through the wizard 3`] = `
"import React from 'react'
import { render } from '@testing-library/react'
import MockComponent from '.'

describe('MockComponent', () => {
  it('creates short description', () => {
    const { asFragment } = render(<MockComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
})
"
`;
